src/OpenTelemetry.AutoInstrumentation.Native/rejit_handler.h

```diff
@@ -90,39 +118,54 @@ class RejitHandlerModule
 class RejitHandler
 {
 private:
+    std::atomic_bool m_shutdown = {false};
+    Lock m_shutdown_lock;
+
     std::mutex m_modules_lock;
     std::unordered_map<ModuleID, std::unique_ptr<RejitHandlerModule>> m_modules;
+    AssemblyProperty* m_pCorAssemblyProperty = nullptr;
 
-    ICorProfilerInfo7* m_profilerInfo7;
+    ICorProfilerInfo7* m_profilerInfo;
+    ICorProfilerInfo10* m_profilerInfo10;
 
-    std::function<HRESULT(RejitHandlerModule*, RejitHandlerModuleMethod*)> m_rewriteCallback;
+    std::shared_ptr<RejitWorkOffloader> m_work_offloader;
+        
+    bool enable_by_ref_instrumentation = false;
+    bool enable_calltarget_state_by_ref = false;
 
     std::mutex m_ngenModules_lock;
     std::vector<ModuleID> m_ngenModules;
 
     void RequestRejitForInlinersInModule(ModuleID moduleId);
-
 public:
-    RejitHandler(ICorProfilerInfo7* pInfo,
-                 std::function<HRESULT(RejitHandlerModule*, RejitHandlerModuleMethod*)> rewriteCallback);
+    RejitHandler(ICorProfilerInfo7* pInfo, std::shared_ptr<RejitWorkOffloader> work_offloader);
+    RejitHandler(ICorProfilerInfo10* pInfo, std::shared_ptr<RejitWorkOffloader> work_offloader);
 
     RejitHandlerModule* GetOrAddModule(ModuleID moduleId);
+    void SetEnableByRefInstrumentation(bool enableByRefInstrumentation);
+    void SetEnableCallTargetStateByRef(bool enableCallTargetStateByRef);
+    bool GetEnableCallTargetStateByRef();
+    bool GetEnableByRefInstrumentation();
 
-    bool TryGetModule(ModuleID moduleId, RejitHandlerModule** moduleHandler);
     void RemoveModule(ModuleID moduleId);
+    bool HasModuleAndMethod(ModuleID moduleId, mdMethodDef methodDef);
 
     void AddNGenModule(ModuleID moduleId);
 
+    void EnqueueForRejit(std::vector<ModuleID>& modulesVector, std::vector<mdMethodDef>& modulesMethodDef);
```

Comment:
This method was removed from code due to #1242
See #1279 PR.

Will it be safe to add it again?

Reply:
This one here seems still important to understand. https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2280#discussion_r1130350462 could be conflicting or both needed to fix the remaining tests.

Reply:
From this native code update, this seems like we are once again susceptible to a race condition with instrumentation when a new assembly is loaded. Let me explain where and how we issue ReJIT requests:
1. At startup we invoke `CorProfiler::InternalAddInstrumentation`, which will initialize the complete list of enabled bytecode instrumentations. We iterate through all loaded modules and do a _blocking_ ReJIT request for instrumentations whose assemblies have been loaded by calling `EnqueueRequestRejitForLoadedModules` with a [promise](https://github.com/zacharycmontoya/opentelemetry-dotnet-instrumentation/blob/c232661aba13dd5ef85468a9504adcff5ff1c673/src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.cpp#L1155-L1156).
2. When a new assembly is loaded we invoke `CorProfiler::ModuleLoadFinished`, which will see if the assembly contains instrumentation(s) and if so it will issue _nonblocking_ ReJIT requests by calling `RequestRejitForLoadedModules` with parameter `enqueueInSameThread=false` (this is required since we can't call ReJIT from the CorProfiler thread).

So number 1 is fine, but number 2 needs to be fixed. All we have to do is call the same API as number 1 and we can block with a certain timeout. I'll create an issue to track this and we can resolve this after this PR is merged.

Reply:
#2452